src/ui/public/vis/vis_types/vislib_vis_type.js

```diff
@@ -23,17 +25,52 @@ export function VislibVisTypeProvider(Private) {
       this._response = esResponse;
       if (this.vis.vislibVis) {
         this.destroy();
-      } else {
-        this.vis.refreshLegend = 0;
       }
 
-      return new Promise((resolve, reject) => {
+      return new Promise(async (resolve, reject) => {
         if (!this._response) return reject();
+
+        let $scope;
+        if (this.vis.params.addLegend) {
+          const legendPositionToVisContainerClassMap = {
+            top: 'vis-container--legend-top',
+            bottom: 'vis-container--legend-bottom',
+            left: 'vis-container--legend-left',
+            right: 'vis-container--legend-right',
+          };
+
+          const getVisContainerClasses = () => {
+            return legendPositionToVisContainerClassMap[this.vis.params.legendPosition];
+          };
+
+          // update the legend class on the parent element
+          $(this.el.parentNode).attr('class', (i, cls) => {
+            return cls.replace(/vis-container--legend-\S+/g, '');
+          }).addClass(getVisContainerClasses());
+
+          $scope = $rootScope.$new();
+          $scope.refreshLegend = 0;
+          $scope.vis = this.vis;
+          $scope.visData = esResponse;
+          $scope.uiState = $scope.vis.getUiState();
+          const legendHtml = $compile('<visualize-legend></visualize-legend>')($scope);
+          this.el.parentNode.appendChild(legendHtml[0]);
```

Comment:
this is terrible as well ... (adding child to the parentNode), however if we don't want to redo css around this atm, i don't see any other way

i'll take a look at redoing the css, maybe it won't be that hard, and then legend can live under visualization-chart